components/raftstore-v2/src/operation/query/local.rs

```diff
@@ -2,44 +2,251 @@
+        }
+    }
+
+    pub async fn snapshot(
+        &mut self,
+        mut req: RaftCmdRequest,
+    ) -> std::result::Result<RegionSnapshot<E::Snapshot>, RaftCmdResponse> {
+        let region_id = req.header.get_ref().region_id;
+        if let Some(snap) = self.try_get_snapshot(req.clone())? {
+            return Ok(snap);
+        }
+
+        if let Some(query_res) = self.try_to_renew_lease(region_id, &req).await? {
+            // If query successful, try again.
+            if query_res.read().is_some() {
+                req.mut_header().set_read_quorum(false);
+                if let Some(snap) = self.try_get_snapshot(req)? {
+                    return Ok(snap);
+                }
+            }
+        }
+
+        let mut err = errorpb::Error::default();
+        err.set_message(format!(
+            "Fail to get snapshot from LocalReader for region {}. Maybe due to `not leader` or `not applied to the current term`",
+            region_id
+        ));
+        let mut resp = RaftCmdResponse::default();
+        resp.mut_header().set_error(err);
+        Err(resp)
+    }
+
+    // try to renew the lease by sending read query where the reading process may
+    // renew the lease
+    async fn try_to_renew_lease(
+        &self,
+        region_id: u64,
+        req: &RaftCmdRequest,
+    ) -> std::result::Result<Option<QueryResult>, RaftCmdResponse> {
+        let (msg, sub) = PeerMsg::raft_query(req.clone());
+        let mut err = errorpb::Error::default();
+        match MsgRouter::send(&self.router, region_id, msg) {
+            Ok(()) => return Ok(sub.result().await),
+            Err(TrySendError::Full(c)) => {
+                TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().reject_reason.channel_full.inc());
+                err.set_message(RAFTSTORE_IS_BUSY.to_owned());
+                err.mut_server_is_busy()
+                    .set_reason(RAFTSTORE_IS_BUSY.to_owned());
+            }
+            Err(TrySendError::Disconnected(c)) => {
+                TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().reject_reason.no_region.inc());
+                err.set_message(format!("region {} is missing", region_id));
+                err.mut_region_not_found().set_region_id(region_id);
+            }
+        }
+
+        let mut resp = RaftCmdResponse::default();
+        resp.mut_header().set_error(err);
+        Err(resp)
+    }
+
+    // If the remote lease will be expired in near future send message
+    // to `raftstore` to renew it
+    fn maybe_renew_lease_in_advance(
+        &self,
+        delegate: &ReadDelegate,
+        req: &RaftCmdRequest,
+        ts: Timespec,
+    ) {
+        if !delegate.need_renew_lease(ts) {
+            return;
+        }
+
+        let region_id = req.header.get_ref().region_id;
+        TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().renew_lease_advance.inc());
+        // Send a read query which may renew the lease
+        let (msg, sub) = PeerMsg::raft_query(req.clone());
```

Comment:
Same as above. If we cache the sub after send successfully, it can reduce the unnecessary renew lease requests.